hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/AddressBookServiceImpl.java

```diff
@@ -231,77 +275,120 @@ private AddressBook parse(FileData fileData) {
     }
 
     /**
-     * Creates a AddressBook object given address book byte array contents. Attempts to convert contents into a valid
-     * NodeAddressBook proto. If successful the NodeAddressBook details including AddressBookEntry entries are extracted
-     * and added to AddressBook object.
+     * Extracts a collection of AddressBookEntry domain objects from NodeAddressBook proto. Sets provided
+     * consensusTimestamp as the consensusTimestamp of each address book entry to ensure mapping to a single
+     * AddressBook
      *
-     * @param fileData fileData object representing address book data
+     * @param nodeAddressBook    node address book proto
+     * @param consensusTimestamp transaction consensusTimestamp
      * @return
      */
-    private AddressBook buildAddressBook(FileData fileData) {
-        long startConsensusTimestamp = fileData.getConsensusTimestamp() + 1;
-        AddressBook.AddressBookBuilder addressBookBuilder = AddressBook.builder()
-                .fileData(fileData.getFileData())
-                .startConsensusTimestamp(startConsensusTimestamp)
-                .fileId(fileData.getEntityId());
-
-        try {
-            NodeAddressBook nodeAddressBook = NodeAddressBook.parseFrom(fileData.getFileData());
-
-            if (nodeAddressBook != null) {
-                if (nodeAddressBook.getNodeAddressCount() > 0) {
-                    addressBookBuilder.nodeCount(nodeAddressBook.getNodeAddressCount());
-                    Collection<AddressBookEntry> addressBookEntryCollection =
-                            retrieveNodeAddressesFromAddressBook(nodeAddressBook, startConsensusTimestamp);
+    private static Collection<AddressBookEntry> retrieveNodeAddressesFromAddressBook(NodeAddressBook nodeAddressBook,
+                                                                                     long consensusTimestamp) throws UnknownHostException {
+        Map<Long, AddressBookEntry> addressBookEntries = new LinkedHashMap<>(); // node id to entry
 
-                    addressBookBuilder.entries((List<AddressBookEntry>) addressBookEntryCollection);
-                }
-            }
-        } catch (Exception e) {
-            log.warn("Unable to parse address book: {}", e.getMessage());
-            return null;
+        for (NodeAddress nodeAddressProto : nodeAddressBook.getNodeAddressList()) {
+            Pair<Long, EntityId> nodeIds = getNodeIds(nodeAddressProto);
```

Comment:
An alternative to this `getNodeIds()` is to just parse the duplicate address book entries regardless then use their extracted nodeID for the map and override with the one in the map if present. Am ok with current approach as well, just noting as I'm not a fan of `Pair`.

Reply:
Yeah, the whole `getLeft()` , `getRight()` makes it illusive.
I'd prefer to return a single object actually.

In that case I could add the following inner class to AddressBookEntry which can serve the purpose

```java
    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class NodeIds {
        private EntityId nodeAccountId;
        private Long nodeId;
    }
```

Then above becomes
```suggestion
            AddressBookEntry.NodeIds nodeIds = getNodeIds(nodeAddressProto);
```
`getLeft()` is replaced with `getNodeAccountId()` and `getRight()` is replaced with `getNodeId()`

Reply:
Decided to leave as it to not add any more churn. It's also something that can be revisited is it comes up again